src/main/java/org/alfresco/heartbeat/HeartBeatJob.java

```diff
@@ -0,0 +1,209 @@
+        final HBDataSenderService hbDataSenderService = (HBDataSenderService) dataMap.get("hbDataSenderService");
+        final JobLockService jobLockService = (JobLockService) dataMap.get("jobLockService");
+
+        if(collector == null)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Exit HeartBeatJob because there is no assigned HB collector");
+            }
+            return;
+        }
+        if(hbDataSenderService == null)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Exit HeartBeatJob because there is no HBDataSenderService");
+            }
+            return;
+        }
+        if(jobLockService == null)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Exit HeartBeatJob because there is no JobLockService");
+            }
+            return;
+        }
+        QName qName = QName.createQName(NamespaceService.SYSTEM_MODEL_1_0_URI, collector.getCollectorId());
+        String lockToken = null;
+        LockCallback lockCallback = new LockCallback(qName);
+        try
+        {
+            // Get a dynamic lock
+            lockToken = acquireLock(lockCallback, qName, jobLockService);
+            collectAndSendDataLocked(collector, hbDataSenderService);
+            // after it finished we want to keep the lock for 30 seconds more
+            try
+            {
+                Thread.sleep(LOCK_TTL_OFFSET);
+            }
+            catch (InterruptedException e)
+            {
+                //
+            }
+        }
+        catch (LockAcquisitionException e)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Skipping collect and send data (could not get lock): " + e.getMessage());
+            }
+        }
+        finally
+        {
+            releaseLock(lockCallback, lockToken, qName, jobLockService);
+        }
+    }
+
+    private void collectAndSendDataLocked(final HBBaseDataCollector collector, final HBDataSenderService hbDataSenderService) throws JobExecutionException
+    {
+        try
+        {
+            List<HBData> data = collector.collectData();
+            try
+            {
+                hbDataSenderService.sendData(data);
+            }
+            catch (Exception e)
+            {
+                logger.warn(e);
+            }
+        }
+        catch (final Exception e)
+        {
+            if (logger.isDebugEnabled())
+            {
+                // Verbose logging
+                logger.debug("Heartbeat job failure from collector: " + collector.getCollectorId(), e);
+            }
+            else
+            {
+                // Heartbeat errors are non-fatal and will show as single line warnings
+                logger.warn(e.toString());
+                throw new JobExecutionException(e);
```

Comment:
It should not depend on the log level whether we throw an exception or not.
Even if debig is enabled, it ios still fine to log it out as a warn. And we should either always or never throw a JobExecutionException in this situation.

Reply:
Well spotted, Stefan.